transports/quic/src/connection/substream.rs

```diff
@@ -19,239 +19,61 @@
+        AsyncWrite::poll_write(Pin::new(&mut self.get_mut().send), cx, buf)
     }
 
-    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
-        let mut inner = self.state.lock();
-
-        let substream_state = inner.unchecked_substream_state(self.id);
-        match substream_state.write_state {
-            WriteState::Open => {}
-            WriteState::Closing => {
-                substream_state.close_waker = Some(cx.waker().clone());
-                return Poll::Pending;
-            }
-            WriteState::Closed => return Poll::Ready(Ok(())),
-            WriteState::Stopped => {
-                let err = quinn_proto::FinishError::Stopped(0u32.into());
-                return Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)));
-            }
-        }
-
-        match inner.connection.send_stream(self.id).finish() {
-            Ok(()) => {
-                let substream_state = inner.unchecked_substream_state(self.id);
-                substream_state.close_waker = Some(cx.waker().clone());
-                substream_state.write_state = WriteState::Closing;
-                Poll::Pending
-            }
-            Err(err @ quinn_proto::FinishError::Stopped(_)) => {
-                Poll::Ready(Err(io::Error::new(io::ErrorKind::ConnectionReset, err)))
-            }
-            Err(quinn_proto::FinishError::UnknownStream) => {
-                // We never make up IDs so the stream must have existed at some point if we get to here.
-                // `UnknownStream` is also emitted in case the stream is already finished, hence just
-                // return `Ok(())` here.
-                Poll::Ready(Ok(()))
-            }
-        }
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        AsyncWrite::poll_flush(Pin::new(&mut self.get_mut().send), cx)
     }
-}
 
-impl Drop for Substream {
-    fn drop(&mut self) {
-        let mut state = self.state.lock();
-        state.substreams.remove(&self.id);
-        // Send `STOP_STREAM` if the remote did not finish the stream yet.
-        // We have to manually check the read stream since we might have
-        // received a `FIN` (without any other stream data) after the last
-        // time we tried to read.
-        let mut is_read_done = false;
-        if let Ok(mut chunks) = state.connection.recv_stream(self.id).read(true) {
-            if let Ok(chunk) = chunks.next(0) {
-                is_read_done = chunk.is_none();
-            }
-            let _ = chunks.finalize();
-        }
-        if !is_read_done {
-            let _ = state.connection.recv_stream(self.id).stop(0u32.into());
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        let this = self.get_mut();
+        if this.closed {
+            // For some reason poll_close needs to be 'fuse'able
```

Comment:
Oh, I wonder where in our stack we call `AsyncWrite::poll_close` after a previous return of `Poll::Ready`. Any idea?

Reply:
I revisited this line, looks like either I was wrong or it was already fixed.